### PR TITLE
Checkpointing/log

### PR DESCRIPTION
**Pull request by jarvislam1999**
Wrote new structs to support incremental game save as well as functions that will potentially be imported by game state next year. Will not be used this year so as long as there are no compilation errors. It should be fine.

### CODE REVIEW

File: src/checkpointing/game.proto [hunk 1]
```diff
@@ -81,3 +81,30 @@ message Game {
...
+    optional bool bool_val = 5;
+}
+
+message Logvalue {
+    optional string tag = 1;
+    optional Logvalueunion value = 2;
+}
+
+message Log {
+    optional string level = 1; // To specify whether it's in game, player, etc
```
Comment (jonahdf) [inline src/checkpointing/game.proto]: We shouldn't be using a string to store data like this. I'd make a union for the object (or whatever we're calling them).

File: src/checkpointing/include/log_save.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+Logvalue *logvalue_init_int(Logvalue *logvalue, int value);
+
+/* Create a new Log struct */
+Log *log_new();
+
+/* Initialize an integer Log struct
+ * Parameters: 
+ * - log: a new log
+ * - level: either "game", "player", "room", "item", etc
+ * - player_id : the id of that level string
```
Comment (jonahdf) [inline src/checkpointing/include/log_save.h]: level_id

File: src/checkpointing/game.proto [hunk 3]
```diff
@@ -81,3 +81,30 @@ message Game {
     optional string curr_player = 6; //ID of player (different from struct pointer of game_state)
     // optional string start_desc = 7;
 }
+
+message Logvalueunion {
```
Comment (jonahdf) [inline src/checkpointing/game.proto]: Should use underscores (style guide)

File: src/checkpointing/game.proto [hunk 4]
```diff
@@ -81,3 +81,30 @@ message Game {
...
+    optional int32 int_val = 1;
+    optional string str_val = 2;
+    optional double double_val = 3;
+    optional int32 char_val = 4; //NO CHAR TYPE SUPPORTED IN .PROTO, SO USING INT INSTEAD
+    optional bool bool_val = 5;
+}
+
+message Logvalue {
```
Comment (jonahdf) [inline src/checkpointing/game.proto]: Should use underscores

File: src/checkpointing/include/log_save.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ * Return: a pointer to that Log
+ */
+Log *log_init_int(Log *log, char *level, char *level_id, char *member, int old_value, int new_value);
+
+/* Create a new Log struct and initialize its value*/
+Logs *logs_new();
+
+/* Add a Log to the array of Log's
+ * Parameters: 
+ * - logs: an existing Logs
```
Comment (jonahdf) [inline src/checkpointing/include/log_save.h]: Say it's an array of logs

File: src/checkpointing/include/log_save.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+ * - log: the Log to add in
+ * Return: a pointer to that Logs
+ */
+Logs *logs_add(Logs *logs, Log *log);
+
+/* Store a changed value information in our array of logs
+ * Parameters: 
+ * - logs: the array of logs
+ * - level: either "game", "player", "room", "item", etc
+ * - player_id : the id of that level string
```
Comment (jonahdf) [inline src/checkpointing/include/log_save.h]: level_id

File: src/checkpointing/include/log_save.h [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ * - member : member of the level string
+ * - old_value: old value of that member
+ * - new_value: new value of that member
+ * Return: 0 if SUCCESS, -1 OTHERWISE
+ */
+int checkpointing_log_int(Logs *logs, char *level, char *level_id, char *member, int old_value, int new_value);
```
Comment (jonahdf) [inline src/checkpointing/include/log_save.h]: What about string, double, or bool?

File: src/checkpointing/src/log_save.c [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  if (log == NULL) {
+    fprintf(stderr, "Need to pass in an initialized Log \n");
+    return NULL;
+  }
+  if (logs->logs_len == 100) {
+    fprintf(stderr, "Logs excedding 100 Log, dump game state\n");
+    return NULL;
+  }
+  logs->logs_len += 1;
+  logs->n_logs += 1;
```
Comment (jonahdf) [inline src/checkpointing/src/log_save.c]: what is n_logs?

File: src/checkpointing/src/log_save.c [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+  logs->logs[logs->logs_len - 1] = log;
```
Comment (jonahdf) [inline src/checkpointing/src/log_save.c]: This is hard to read—shouldn't we be avoiding using the same exact variable name for two different levels?

File: src/checkpointing/game.proto (hunk 3 above)
Comment (jarvislam1999) [reply src/checkpointing/game.proto]: Protobuf doesn't do well with underscores for some reasons, so I will keep it out for now

File: src/checkpointing/game.proto (hunk 4 above)
Comment (jarvislam1999) [reply src/checkpointing/game.proto]: Protobuf doesn't do well with underscores for some reasons, so I will keep it out for now

File: src/checkpointing/game.proto (hunk 1 above)
Comment (jarvislam1999) [reply src/checkpointing/game.proto]: Well the string here is just store an enum, not the type itself

File: src/checkpointing/include/log_save.h (hunk 2 above)
Comment (jarvislam1999) [reply src/checkpointing/include/log_save.h]: Good catch

File: src/checkpointing/include/log_save.h (hunk 5 above)
Comment (jarvislam1999) [reply src/checkpointing/include/log_save.h]: Done

File: src/checkpointing/include/log_save.h (hunk 6 above)
Comment (jarvislam1999) [reply src/checkpointing/include/log_save.h]: Done

File: src/checkpointing/include/log_save.h (hunk 7 above)
Comment (jarvislam1999) [reply src/checkpointing/include/log_save.h]: It's just int for now, future gen can decide whether to opt for more versions or use void

File: src/checkpointing/src/log_save.c (hunk 8 above)
Comment (jarvislam1999) [reply src/checkpointing/src/log_save.c]: n_logs is a hidden variable in protobuf that needs to be assigned value to work

File: src/checkpointing/src/log_save.c (hunk 9 above)
Comment (jarvislam1999) [reply src/checkpointing/src/log_save.c]: Changed logs to l

**Pull request merged**